common/remote_util.cc

```diff
@@ -77,6 +98,33 @@ absl::Status RemoteUtil::Scp(std::vector<std::string> source_filepaths,
   return process_factory_->Run(start_info);
 }
 
+absl::Status RemoteUtil::Sftp(const std::string& commands,
+                              const std::string& initial_local_dir,
+                              bool compress) {
+  // sftp doesn't take |commands| as argument, so write it to a temp file.
+  std::string cmd_path =
+      path::Join(path::GetTempDir(), "__sftp_cmd__" + Util::GenerateUniqueId());
+  RETURN_IF_ERROR(path::WriteFile(cmd_path, commands),
+                  "Failed to write sftp commands to '%s'", cmd_path);
+
+  // -p preserves timestamps. This enables timestamp-based up-to-date checks.
+  ProcessStartInfo start_info;
+  start_info.flags = ProcessFlags::kNoWindow;
+  start_info.command = absl::StrFormat(
+      "%s %s %s -p -b %s %s", sftp_command_,
+      quiet_ || verbosity_ < 2 ? "-q" : "", compress ? "-C" : "",
+      QuoteForWindows(cmd_path), QuoteForWindows(user_host_));
+  start_info.name = "sftp";
+  start_info.startup_dir = initial_local_dir;
+  start_info.forward_output_to_log = forward_output_to_log_;
+
+  RETURN_IF_ERROR(process_factory_->Run(start_info));
```

Comment:
Do you want to include `cmd_path` in the returned error?

Reply:
Currently users are supposed to run with -vv to see the full commands, so that you can repro them in a cmd window. But that requires reading the README. There's an open issue about making error messages more useful in general. I think printing out full commands in case of an error makes sense, but I'd like to do that in a separate CL for all commands.